File: src/class037/C5_ValidateBinarySearchTree.py
class TreeNode:
    """树节点定义"""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 全局变量定义
MAXN = 10001
stack = [None] * MAXN
r = 0
min_val = float('inf')
max_val = float('-inf')

def isValidBST1(head):
    """使用中序遍历的迭代方式检查二叉搜索树"""
    global r, stack
    if head is None:
        return True
    pre = None
    r = 0
    while r > 0 or head is not None:
        if head is not None:
            stack[r] = head
            r += 1
            head = head.left
        else:
            head = stack[r - 1]
            r -= 1
            if pre is not None and pre.val >= head.val:
                return False
            pre = head
            head = head.right
    return True

def isValidBST2(head):
    """使用递归方式检查二叉搜索树，并更新最小值和最大值"""
    global min_val, max_val
    if head is None:
        min_val = float('inf')
        max_val = float('-inf')
        return True
    lok = isValidBST2(head.left)
    lmin = min_val
    lmax = max_val
    rok = isValidBST2(head.right)
    rmin = min_val
    rmax = max_val
    min_val = min(min(lmin, rmin), head.val)
    max_val = max(max(lmax, rmax), head.val)
    return lok and rok and lmax < head.val and head.val < rmin

# 注意：在实际提交代码时需要将函数名从 isValidBST1 和 isValidBST2 修改为 isValidBST

File: src/class037/C5_ValidateBinarySearchTree.cpp
#include <climits> // For INT_MAX and INT_MIN
#include <vector>
using namespace std;

// 二叉树节点定义
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
};

class Solution {
    // 定义堆栈存储树的节点，最大容量为10001
    static const int MAXN = 10001;
    vector<TreeNode*> stack;
    int r;

    // 检验二叉搜索树方法1
    bool isValidBST1(TreeNode* head) {
        if (head == nullptr) {
            return true;
        }
        TreeNode* pre = nullptr;
        r = 0;
        stack.resize(MAXN);
        while (r > 0 || head != nullptr) {
            if (head != nullptr) {
                stack[r++] = head;
                head = head->left;
            } else {
                head = stack[--r];
                if (pre != nullptr && pre->val >= head->val) {
                    return false;
                }
                pre = head;
                head = head->right;
            }
        }
        return true;
    }

    // 用于存储子树的最小值和最大值
    long minn, maxx;

    // 检验二叉搜索树方法2
    bool isValidBST2(TreeNode* head) {
        if (head == nullptr) {
            minn = LONG_MAX;
            maxx = LONG_MIN;
            return true;
        }
        bool lok = isValidBST2(head->left);
        long lmin = minn;
        long lmax = maxx;
        bool rok = isValidBST2(head->right);
        long rmin = minn;
        long rmax = maxx;
        minn = min({lmin, rmin, head->val});
        maxx = max({lmax, rmax, head->val});
        return lok && rok && lmax < head->val && head->val < rmin;
    }
};
